### PR TITLE
Add EventStore Hosting and Client integrations

### LINKED ISSUE #236: EventStore Integration
### .NET Aspire issue link

_No response_

### Overview

EventStoreDB is the event-native database, where business events are immutably stored and streamed. Designed for event-sourced, event-driven, and microservices architecture.

### Usage example

```csharp
var eventStore = builder.AddEventStore("eventstore")
    .WithDataVolume();
```

### Additional context

_No response_

### Help us help you

Yes, I'd like to be assigned to work on this item

### PR DESCRIPTION
**Pull request by fredimachado**
**Closes #236**

This pull request includes several updates to the Aspire CommunityToolkit including new projects for EventStore Hosting and Client integrations, a simple "bank account" example project, tests and updates to dependencies.

### Project Additions and Updates:
Added new EventStore projects to the solution, including `Aspire.CommunityToolkit.Hosting.EventStore`, `Aspire.CommunityToolkit.EventStore`, and their respective test and example projects.

### Dependency Updates:
Added new dependencies in `Directory.Packages.props`, including `AspNetCore.HealthChecks.EventStore.gRPC`, `EventStore.Client.Extensions.OpenTelemetry`, and `EventStore.Client.Grpc.Streams`. (`Directory.Packages.props`) [[1]](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-5baf5f9e448ad54ab25a091adee0da05d4d228481c9200518fcb1b53a65d4156)

### Code Ownership:
Updated the CODEOWNERS file to include new EventStore projects and assigned ownership to @fredimachado. (CODEOWNERS)

### New Example Project:
Added a simple "bank account" example project for EventStore service. (examples/eventstore/Aspire.CommunityToolkit.Hosting.EventStore.ApiService) [[1]](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-979a3162ee391e64220c83e4cc44fd42a4088f5f4e481a608cbbe7dab5110a04)

## PR Checklist

- [x] Created a feature/dev branch in your fork (vs. submitting directly from a commit on main)
- [x] Based off latest main branch of toolkit
- [x] PR doesn't include merge commits (always rebase on top of our main, if needed)
- [x] New integration
  - [x] Docs are written
  - [x] Added description of major feature to project description for NuGet package (4000 total character limit, so don't push entire description over that)
- [x] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] Contains **NO** breaking changes
- [x] Every new API (including internal ones) has full XML docs
- [x] Code follows all style conventions

## Other information

Docs PR: https://github.com/dotnet/docs-aspire/pull/2143

### CODE REVIEW

File: CODEOWNERS
Comment (aaronpowell) [inline CODEOWNERS:1]: I'll add you as a contributor to address the errors reported here - assuming you are wanting to come onboard as a maintainer.

File: src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs
Comment (aaronpowell) [inline src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:44]:
```suggestion
    public static IResourceBuilder<EventStoreResource> AddEventStore(this IDistributedApplicationBuilder builder, [ResourceName] string name, int? port = null)
```

Let's avoid arbitrary line breaks.

Also added the `ResourceName` attribute to the name parameter for the analyzer

File: src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs
Comment (aaronpowell) [inline src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:65]: I'd prefer that we pass the resource into the constructor and then resolve the connection string when the health check is executed.

If the connection string can't be resolved, the resource is unhealthy.

This way we're avoiding relying on the event and we also don't have to use the null forgiving operator.

File: src/CommunityToolkit.Aspire.Hosting.EventStore/README.md
Comment (aaronpowell) [inline src/CommunityToolkit.Aspire.Hosting.EventStore/README.md:9]:
```suggestion
In your app host project, install the package using the following command:
```

File: src/CommunityToolkit.Aspire.Hosting.EventStore/README.md
Comment (aaronpowell) [inline src/CommunityToolkit.Aspire.Hosting.EventStore/README.md:3]:
```suggestion
Provides extension methods and resource definitions for the .NET Aspire app host to support running [EventStore](https://www.eventstore.com) containers.
```

File: src/CommunityToolkit.Aspire.Hosting.EventStore/README.md
Comment (aaronpowell) [inline src/CommunityToolkit.Aspire.Hosting.EventStore/README.md:17]:
```suggestion
Then, in the _Program.cs_ file of app host, add a EventStore resource and consume the connection using the following methods:
```

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (aaronpowell) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:20]: it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (aaronpowell) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:35]: it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

File: src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs
Comment (aaronpowell) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:65]: Actually, is that a limitation of the external health check package?

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (aaronpowell) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:59]: it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (aaronpowell) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs:86]: it's better to use the method that waits for the resource to be healthy, so consideration is taken for the app in the container starting up

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStorePublicApiTests.cs
Comment (aaronpowell) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStorePublicApiTests.cs:1]: can we also get some tests that the right annotations are added for the volume/data bind mounts

**Review by aaronpowell — CHANGES_REQUESTED** [review]
I don't see tests for the client integration, only the hosting, I assume that they are still to come

File: CODEOWNERS
Comment (fredimachado) [reply CODEOWNERS:1]: Sure, I'd love to help :)

File: src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs
Comment (fredimachado) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:65]: Yes,. Unfortunately, it only takes a connection string at the moment.

Comment (aaronpowell) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:65]: All good, I'd forgotten you were using an external library for the health check on initial review (when I wrote the first comment) and then realised at the end.

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStorePublicApiTests.cs
Comment (fredimachado) [reply tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStorePublicApiTests.cs:1]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (fredimachado) [reply tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (fredimachado) [reply tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (fredimachado) [reply tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs
Comment (fredimachado) [reply tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/AppHostTests.cs]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: src/CommunityToolkit.Aspire.Hosting.EventStore/README.md
Comment (fredimachado) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/README.md]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: src/CommunityToolkit.Aspire.Hosting.EventStore/README.md
Comment (fredimachado) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/README.md]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs
Comment (fredimachado) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

File: src/CommunityToolkit.Aspire.Hosting.EventStore/README.md
Comment (fredimachado) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/README.md]: Addressed in the last commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/7deb7d5bc32aa830d8266342076c010509f75b88

**Comment by fredimachado** [comment]
> I don't see tests for the client integration, only the hosting, I assume that they are still to come

Addressed in this commit: https://github.com/CommunityToolkit/Aspire/pull/277/commits/8fc728b009144a6e1db5fc155801875268a6d550

**Review by aaronpowell — APPROVED** [review]
Looks really good @fredimachado 

I've kicked off a build, once it's clean we can look to merge it in.

You should also have an invite to the contributor team (which will sort out the CODEOWNERS)

File: src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs
Comment (Alirexaa) [inline src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:166]: Why we have seperate method for this instead of using this in `WithDataVolume`?

Comment (fredimachado) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:166]: This one is for mounting logs (`/var/logs/eventstore`) while the other one mounts data (`/var/lib/eventstore`).
Should we combine both? I usually don't store logs as it can get pretty big.

Comment (Alirexaa) [reply src/CommunityToolkit.Aspire.Hosting.EventStore/EventStoreBuilderExtensions.cs:166]: If users usually don't store logs that should be fine.

File: src/CommunityToolkit.Aspire.EventStore/AspireEventStoreExtensions.cs
Comment (fredimachado) [reply src/CommunityToolkit.Aspire.EventStore/AspireEventStoreExtensions.cs]: Addressed in this commit https://github.com/CommunityToolkit/Aspire/pull/277/commits/9bb0b5d6e6b3769b21791931f27d077aabc8fef3

File: tests/CommunityToolkit.Aspire.EventStore.Tests/ConformanceTests.cs
Comment (Alirexaa) [inline tests/CommunityToolkit.Aspire.EventStore.Tests/ConformanceTests.cs:72]:
```suggestion
        
```

**Review by Alirexaa — APPROVED** [review]
Besides of suggested change totally LGTM.

**Review by Alirexaa — CHANGES_REQUESTED** [review]
I reviewed tests and did not see tests around persistence and health check functionality.

Could you add these tests?
these are some examples at 
https://github.com/CommunityToolkit/Aspire/blob/main/tests/CommunityToolkit.Aspire.Hosting.Meilisearch/MeilisearchFunctionalTests.cs

**Comment by fredimachado** [comment]
> I reviewed tests and did not see tests around persistence and health check functionality.
> 
> Could you add these tests? there are some examples at https://github.com/CommunityToolkit/Aspire/blob/main/tests/CommunityToolkit.Aspire.Hosting.Meilisearch/MeilisearchFunctionalTests.cs

I thought [AppHostTests](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-39c20f8f153ec2acf9f9d92e8ee70f4c7d74c9b6940d22651514c6dbbe67c82a) would also count as persistence tests, since it uses the example API to append and then read events from EventStore.

**Comment by fredimachado** [comment]
> > > I reviewed tests and did not see tests around persistence and health check functionality.
> > > Could you add these tests? there are some examples at https://github.com/CommunityToolkit/Aspire/blob/main/tests/CommunityToolkit.Aspire.Hosting.Meilisearch/MeilisearchFunctionalTests.cs
> > 
> > 
> > I thought [AppHostTests](https://github.com/CommunityToolkit/Aspire/pull/277/files#diff-39c20f8f153ec2acf9f9d92e8ee70f4c7d74c9b6940d22651514c6dbbe67c82a) would also count as persistence tests, since it uses the example API to append and then read events from EventStore.
> 
> AppHost does not stop across tests. So it's not enough for all cases.

All good. I added the functional tests based on the Meilisearch ones. https://github.com/CommunityToolkit/Aspire/pull/277/commits/d78ff8fa9122e4442452bf77cfbe752df92804b8

**Review by aaronpowell — APPROVED** [review]
I'm happy with how it's looking with the extra tests, but I'll defer to @Alirexaa for final say

**Comment by fredimachado** [comment]
@Alirexaa I removed the Log volume/mount since it's probably not going to be used often, and we can always use `WithVolume` extension method in the ResourceBuilder to manually add the log volume if necessary.

**Comment by fredimachado** [comment]
@aaronpowell I noticed the SWA example test is stuck during the build's Test step.
I noticed some packages have vulnerabilities, not sure if it has something to do with that. Windows Tests are runnng fine though.
```
Run npm install -g @azure/static-web-apps-cli
(node:3792) ExperimentalWarning: CommonJS module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm warn deprecated rimraf@2.7.1: Rimraf versions prior to v4 are no longer supported

changed 345 packages in 15s

72 packages are looking for funding
  run `npm fund` for details
(node:3822) ExperimentalWarning: CommonJS module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/debug/src/node.js is loading ES Module /opt/hostedtoolcache/node/23.3.0/x64/lib/node_modules/npm/node_modules/supports-color/index.js using require().
Support for loading ES Module in require() is an experimental feature and might change at any time
(Use `node --trace-warnings ...` to show where the warning was created)

added 199 packages, and audited [20](https://github.com/CommunityToolkit/Aspire/actions/runs/12020126580/job/33508109977?pr=277#step:18:21)0 packages in 3s

42 packages are looking for funding
  run `npm fund` for details

2 vulnerabilities (1 moderate, 1 high)

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
```

**Comment by aaronpowell** [comment]
Yeah, unfortunately that's out of our control, we're installing the latest package version

**Comment by aaronpowell** [comment]
It looks like we have a test that is hanging and causing the Linux runner to time out after several hours.

Assuming that these tests pass locally, this is going to be a fun one to try and resolve 🤣. Comparing the test results artifacts between Windows and Linux the only missing TRX file is the `CommunityToolkit.Aspire.Hosting.EventStore.Tests`, which suggests that the problematic tests are in that project.

The most likely candidates for the issue is going to be the tests in the `EventStoreFunctionalTests.cs` file. When I've seen this problem in the past it is related to the test waiting for something to complete that never completes but is still running "successfully" so a timeout on the async method doesn't work.

Here's the steps I'd go through to try and resolve it:

1. Add a `Skip` attribute to all tests in that file and push
  1.1. If the runner still times out, also skip the test that use the example project - basically all tests that use Docker since they don't run on Windows
3. Add one test back per push until you find the test that causes the runner to not exit

Once you know what test is the problematic one, we'll be able to help diagnose the underlying problem.

**Comment by fredimachado** [comment]
All good, thanks for the suggestions. I'll try it.

**Comment by aaronpowell** [comment]
> All good, thanks for the suggestions. I'll try it.

Unfortunately, this isn't the first time I've seen runners hang 🤣

**Comment by aaronpowell** [comment]
Well, looks like we found where the tests are that are the problem

**Comment by fredimachado** [comment]
It seems the we had to change permissions for the data bind mount to work on the Ubuntu build.
cc @aaronpowell @Alirexaa

**Comment by aaronpowell** [comment]
> It seems the we had to change permissions for the data bind mount to work on the Ubuntu build. cc @aaronpowell @Alirexaa
> 
> image

That's interesting, wonder why...

**Comment by aaronpowell** [comment]
Just waiting for another review from @Alirexaa and it should be good to go

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs
Comment (Alirexaa) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs:135]: This could incur flakiness.

Related https://github.com/dotnet/aspire/issues/4878

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs
Comment (Alirexaa) [inline tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs:207]: Is this a light (meaning small image size and fast startup) container image?
If it is we could use this image for dependent resources in health check tests like this test to speed up our tests.

**Review by Alirexaa — APPROVED** [review]
@fredimachado Please take a look at my comments.
Otherwise LGTM

**Comment by Alirexaa** [comment]
@aaronpowell, CodeQL is blocking me to merge PR.

**Comment by aaronpowell** [comment]
> @aaronpowell, CodeQL is blocking me to merge PR.

Grr, I need to find out why that happens

File: tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs
Comment (fredimachado) [reply tests/CommunityToolkit.Aspire.Hosting.EventStore.Tests/EventStoreFunctionalTests.cs:207]: Exactly, I used this specifically because it's lightweight and fast to start.
I've seen usages of this image for the same purpose in the main Aspire repo.

**Comment by fredimachado** [comment]
Thanks @aaronpowell and @Alirexaa for your support.

**Pull request merged**
